contrib: Add ContainerLab-based BGP CPlane development environment (#28292)
cilium-team/cilium: Documentation/contributing/development/bgp_cplane.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _bgp_cplane_contributing:

=================
BGP Control Plane
=================

This section is specific to BGP Control Plane contributions.

Development Environment
=======================

> BGP Control Plane requires BGP peer for testing. You can use `ContainerLab <https://containerlab.dev/>`_ and `Kind <https://kind.sigs.k8s.io/docs/user/configuration/>`_-based development environment. The following diagram shows the topology.

Reviewer: Couple of suggestions here for grammar and word choice.
```suggestion
BGP Control Plane requires a BGP peer for testing. This section describes a `ContainerLab <https://containerlab.dev/>`_ and `Kind <https://kind.sigs.k8s.io/docs/user/configuration/>`_-based development environment. The following diagram shows the topology:
```

Reviewer: Additionally, can you modify these links so they are block-level hyperlink targets? See https://docs.cilium.io/en/stable/contributing/docs/docsstyle/#links

.. image:: _static/bgp-cplane-dev-topo.png
    :align: center

> Role of each nodes are following

Reviewer: I'm not sure if the grammar is right here, I left a suggestion for rewording this.

```suggestion
The following describes the role of each node:
```

> * ``router0`` is an `FRRouting (FRR) <https://frrouting.org/>`_ router. It is pre-configured with a minimal peering settings with server0 and server1.

Reviewer: The "a" tells the reader that the noun it applies to is singular, but "settings" is plural. I'd suggest the following:

```suggestion
* ``router0`` is an `FRRouting (FRR) <https://frrouting.org/>`_ router. It is pre-configured with minimal peering settings with server0 and server1.
```

> * ``server0`` and ``server1`` are Kind nodes (technically, they are ``nicolaka/netshoot`` containers that share network namespaces with Kind nodes).

Reviewer: Do you think you could reword this so there aren't parentheses used here? This is really cool that these are netshoot containers that share a netns with kind nodes, it shows off the power and usefulness of containerlab. I think it makes sense to highlight that. Here's an example of what I'm thinking:

```suggestion
* ``server0`` and ``server1`` are ``nicolaka/netshoot`` containers that each share a network namespace with their own Kind node.
```

* ``server2`` is a non-Cilium ``nicolaka/netshoot`` node useful for testing traffic connectivity from outside of the k8s cluster.

Deploy Lab
----------

.. code-block:: shell-session

   $ make kind-bgp-v4

.. note::
        Above example setups IPv4 single-stack environment. You can change ``v4`` part to ``v6`` or ``dual`` to setup IPv6 single-stack or dual-stack environment respectively. Same for all below examples.

Install Cilium on the Lab
-------------------------

> Install Cilium on the lab with your favorite way. Following example assumes you are modifying the source and want to build your image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. You can add Helm values to deploy BGP Control Plane with different Cilium configuration.

Reviewer: There are a couple of changes that I'd recommend here for grammar and word choice:

```suggestion
Install Cilium on the lab with your favorite way. The following example assumes you are modifying the source and want to build your own image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. If needed, you can add Helm values to deploy BGP Control Plane with a different Cilium configuration.
```

.. code-block:: shell-session

   $ make docker-cilium-image docker-operator-generic-image
   $ kind --name bgp-cplane-dev-v4 load docker-image quay.io/cilium/cilium:latest quay.io/cilium/operator-generic:latest
   $ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml --helm-set image.override="quay.io/cilium/cilium:latest" --helm-set image.pullPolicy=Never --helm-set operator.image.override="quay.io/cilium/operator-generic:latest" --helm-set operator.image.pullPolicy=Never

Peering with Router
-------------------

> Peering Cilium nodes with FRR by applying CiliumBGPPeeringPolicy.

Reviewer: A couple of grammar suggestions to reword this as an instruction for the user:

```suggestion
Peer Cilium nodes with FRR by applying a CiliumBGPPeeringPolicy:
```

.. code-block:: shell-session

   $ make kind-bgp-v4-apply-policy

.. note::
        At this point, there are only minimal peering settings on the policy and no advertisement configuration present. You need to edit policies, for example, with ``kubectl edit bgpp`` to realize your desired settings. If you need to change the router side, you can edit FRRouting settings with ``docker exec -it clab-bgp-cplane-dev-v4-router0 vtysh``.

Validating Peering Status
-------------------------

.. code-block:: shell-session

   $ cilium bgp peers

Validating Connectivity to Cilium Nodes from Non-Cilium Node
------------------------------------------------------------

.. code-block:: shell-session

   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.1.2

>    $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.2.2

Reviewer: Is there any output here that the user should expect if the connectivity is successful? Or is there no output and a zero RC?

Author: Good point. Updated with the expected output and description for it.

.. only:: not (epub or latex or html)

> 

Reviewer 2: Adding a comment here since I can't comment on the png file. Consider adding the Kind node name to the diagram, e.g. `server0 (Kind control-plane Node, Cilium)`.

Author: Good idea! I added the control-plane and worker, respectively.

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _bgp_cplane_contributing:

=================
BGP Control Plane
=================

> This section is specific to BGP Control Plane contributions.

Reviewer 2: nit: consider linking `BGP Control Plane` to the BGP CP [docs](https://docs.cilium.io/en/stable/network/bgp-control-plane/).

Author: Good idea. Done.

Development Environment
=======================

BGP Control Plane requires a BGP peer for testing. This section describes a `ContainerLab`_ and `Kind`_-based development environment. The following diagram shows the topology:

.. _ContainerLab: https://containerlab.dev/
.. _Kind: https://kind.sigs.k8s.io/

.. image:: _static/bgp-cplane-dev-topo.png
    :align: center

The following describes the role of each node:

* ``router0`` is an `FRRouting (FRR)`_ router. It is pre-configured with minimal peering settings with server0 and server1.
* ``server0`` and ``server1`` are ``nicolaka/netshoot`` containers that each share a network namespace with their own Kind node.
* ``server2`` is a non-Cilium ``nicolaka/netshoot`` node useful for testing traffic connectivity from outside of the k8s cluster.

.. _FRRouting (FRR): https://frrouting.org/

Deploy Lab
----------

.. code-block:: shell-session

   $ make kind-bgp-v4

.. note::

>         Above example setups IPv4 single-stack environment. You can change ``v4`` part to ``v6`` or ``dual`` to setup IPv6 single-stack or dual-stack environment respectively. Same for all below examples.

Reviewer: Can you replace the use of "Above" and "below" here? Sometimes translators have a hard time with these words in this context. Using "prior" and "following" are great alternatives.

Additionally, because "setup" is used as a verb here, it needs to be separated into two words. It's something that I had to learn when reviewing this :). This page is really helpful: https://www.grammarbook.com/blog/spelling/setup-vs-set-up/.

I also left some suggestions for grammar.

```suggestion
        The prior example sets up an IPv4 single-stack environment. You can change the ``v4`` part to ``v6`` or ``dual`` to set up an IPv6 single-stack or dual-stack environment respectively. The same goes for the following examples.
```

Author: Thanks for the input! I didn't know they were different words 😮 

Applied your suggestion.

Install Cilium on the Lab
-------------------------

Install Cilium on the lab with your favorite way. The following example assumes you are modifying the source and want to build your own image. The minimal mandatory Helm values are provided in ``contrib/containerlab/bgp-cplane-dev-v4/values.yaml``. If needed, you can add Helm values to deploy BGP Control Plane with a different Cilium configuration.

.. code-block:: shell-session

   $ make docker-cilium-image docker-operator-generic-image
   $ kind --name bgp-cplane-dev-v4 load docker-image quay.io/cilium/cilium:latest quay.io/cilium/operator-generic:latest

>    $ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml --helm-set image.override="quay.io/cilium/cilium:latest" --helm-set image.pullPolicy=Never --helm-set operator.image.override="quay.io/cilium/operator-generic:latest" --helm-set operator.image.pullPolicy=Never

Reviewer 2: This workflow could be simplified by adding the ability to specify the kind cluster name. For example:

```sh
$ git diff Makefile
diff --git a/Makefile b/Makefile
index 3730ee6c01..bd0d692cb0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,8 @@
 # Copyright Authors of Cilium
 # SPDX-License-Identifier: Apache-2.0

+CLUSTER_NAME ?= kind
+
 ##@ Default
 all: precheck build postcheck ## Default make target that perform precheck -> build -> postcheck
        @echo "Build finished."
@@ -576,7 +578,7 @@ kind-build-image-agent: ## Build cilium-dev docker image

 $(eval $(call KIND_ENV,kind-image-agent))
 kind-image-agent: kind-ready kind-build-image-agent ## Build cilium-dev docker image and import it into kind.
-       $(QUIET)kind load docker-image $(LOCAL_AGENT_IMAGE)
+       $(QUIET)kind load docker-image $(LOCAL_AGENT_IMAGE) -n $(CLUSTER_NAME)

 $(eval $(call KIND_ENV,kind-build-image-operator))
 kind-build-image-operator: ## Build cilium-operator-dev docker image
@@ -584,7 +586,7 @@ kind-build-image-operator: ## Build cilium-operator-dev docker image

 $(eval $(call KIND_ENV,kind-image-operator))
 kind-image-operator: kind-ready kind-build-image-operator ## Build cilium-operator-dev docker image and import it into kind.
-       $(QUIET)kind load docker-image $(LOCAL_OPERATOR_IMAGE)
+       $(QUIET)kind load docker-image $(LOCAL_OPERATOR_IMAGE) -n $(CLUSTER_NAME)

 $(eval $(call KIND_ENV,kind-build-clustermesh-apiserver))
 kind-build-clustermesh-apiserver: ## Build cilium-clustermesh-apiserver docker image
```

A user would then:

```sh
$ CLUSTER_NAME=bgp-cplane-dev-v4 make kind-image
$ cilium install --chart-directory install/kubernetes/cilium -f contrib/containerlab/bgp-cplane-dev-v4/values.yaml
```

Author: Following the workflow of the `make kind-xxx` target sounds good to me. I put your suggestion into a separate commit and updated the document. Please take a look.

Peering with Router
-------------------

Peer Cilium nodes with FRR by applying a CiliumBGPPeeringPolicy:

.. code-block:: shell-session

   $ make kind-bgp-v4-apply-policy

.. note::
        At this point, there are only minimal peering settings on the policy and no advertisement configuration present. You need to edit policies, for example, with ``kubectl edit bgpp`` to realize your desired settings. If you need to change the router side, you can edit FRRouting settings with ``docker exec -it clab-bgp-cplane-dev-v4-router0 vtysh``.

Validating Peering Status
-------------------------

.. code-block:: shell-session

>    $ cilium bgp peers

Reviewer 2: Consider adding the expected response from the command or add a short note that states the expected session state, routers received/advertised, etc.

Reviewer: +1 from me, that's a great idea.

Author: Thanks, I added an expected result and description.

Validating Connectivity to Cilium Nodes from Non-Cilium Node
------------------------------------------------------------

> The below example validates connectivity from server2 to server0 (10.0.1.2) and server1 (10.0.2.2). You should see the packet goes through router0 (10.0.3.1).

Reviewer: I think this needs to be "go" here instead of "goes"

```suggestion
The below example validates connectivity from server2 to server0 (10.0.1.2) and server1 (10.0.2.2). You should see the packet go through router0 (10.0.3.1).
```

Author: The `packet` needed to be `packets` as well. Fixed.

.. code-block:: shell-session

   $ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.1.2 -r
   Start: 2023-10-04T02:24:58+0000
   HOST: server2                     Loss%   Snt   Last   Avg  Best  Wrst StDev

>    1.|-- 10.0.3.1                   0.0%    10    0.0   0.1   0.0   0.1   0.0

Reviewer: I'm not familiar with the expected output here, but is there some whitespace misalignment here? The data in the columns don't seem to line up with their associated headers. For instance, I would have expected:

```
$ docker exec -it clab-bgp-cplane-dev-v4-server2 mtr 10.0.2.2 -r
Start: 2023-10-04T02:25:58+0000
HOST: server2                     Loss%   Snt   Last   Avg  Best  Wrst StDev
1.|-- 10.0.3.1                     0.0%    10    0.1   0.1   0.0   0.1   0.0
2.|-- 10.0.2.2                     0.0%    10    0.1   0.1   0.0   0.2   0.0
```

The same goes for the table in the second command.

Sometimes weird stuff can happen when copying and pasting command output, so I just wanted to double check.

Author: Yes, as you said, something wrong happened when I copied and pasted the output. I fixed it.
